packages/grafana-ui/src/components/Monaco/CodeEditor.tsx

```diff
@@ -1,35 +1,113 @@
 import React from 'react';
 import { withTheme } from '../../themes';
 import { Themeable } from '../../types';
-import { KeyCode, editor, KeyMod } from 'monaco-editor/esm/vs/editor/editor.api';
+import { CodeEditorProps } from './types';
+import * as monaco from 'monaco-editor/esm/vs/editor/editor.api';
 import ReactMonaco from 'react-monaco-editor';
 
-export interface CodeEditorProps {
-  value: string;
-  language: string;
-  width?: number | string;
-  height?: number | string;
-
-  readOnly?: boolean;
-  showMiniMap?: boolean;
+function createDependencyProposals(range: any): monaco.languages.CompletionItem[] {
+  // returning a static list of proposals, not even looking at the prefix (filtering is done by the Monaco editor),
+  // here you could do a server side lookup
+  return [
+    {
+      label: '${__field.name} (value)',
+      kind: monaco.languages.CompletionItemKind.Value,
+      documentation: 'The field value name (...)',
+      insertText: '${__field.name}',
+      detail: 'Keword detail',
+      range: range,
+    },
+    {
+      label: '${__field.name} (Text)',
+      kind: monaco.languages.CompletionItemKind.Text,
+      documentation: 'The field value name (...)',
+      insertText: '${__field.name}',
+      range: range,
+    },
+    {
+      label: '${__field.name} (Field)',
+      kind: monaco.languages.CompletionItemKind.Field,
+      documentation: 'The field value name (...)',
+      insertText: '${__field.name}',
+      detail: 'Keword detail',
+      range: range,
+    },
+    {
+      label: '${__field.name} (Property)',
+      kind: monaco.languages.CompletionItemKind.Property,
+      documentation: 'The field value name (...)',
+      insertText: '${__field.name}',
+      detail: 'Keword detail',
+      range: range,
+    },
+    {
+      label: '${__field.name} (Function)',
+      kind: monaco.languages.CompletionItemKind.Function,
+      documentation: 'The field value name (...)',
+      insertText: '${__field.name}',
+      range: range,
+      detail: 'Keword detail',
+    },
+  ];
+}
+function registerXXXX(language: string): monaco.IDisposable {
+  console.log('registerCompletionItemProvider', language);
+  return monaco.languages.registerCompletionItemProvider(language, {
```

Comment:
sadly -- it looks like completion needs to be added globally to the language :(  would be much nicer if it was linked to the editor instance.

It does look like you can register and destory the instances, but this has the not great side affect that really one suggest context is active for each language at a time.

NOTE: this works on top of the standard language completions :tada: